### PR TITLE
Add EmbeddedFormViewController

### PR DESCRIPTION
**Pull request by porter-stripe**
## Summary
- Adds EmbeddedFormViewController which will be presented when we need to collect user input 
- Adds snapshot tests
- Following PR will present this VC when a row is tapped and we need to collect user input
- Started with a copy of the PaymentSheetVerticalViewController then trimmed it down

## Motivation
- Embedded

## Testing
- Snapshots

## Changelog
N/A

### CODE REVIEW

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+            callToAction: callToAction,
+            animated: true
+        )
+    }
+
+    func updateMandate() {
+        let mandateProvider = VerticalListMandateProvider(configuration: configuration, elementsSession: elementsSession, intent: intent)
+        let newMandateText = mandateProvider.mandate(for: selectedPaymentOption?.paymentMethodType,
+                                                     savedPaymentMethod: selectedPaymentOption?.savedPaymentMethod,
+                                                     bottomNoticeAttributedString: paymentMethodFormViewController.bottomNoticeAttributedString)
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift:210]: nit I've realized that Xcode's indentation style is different from this, so let's try to use that.  You can let it auto-format by highlighting code and hitting ctrl-m

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        if !didSendLogShow {
+            // Only send this once to match the behavior of horizontal mode
+            didSendLogShow = true
+            analyticsHelper.logShow(showingSavedPMList: false) // We never show the saved PM list first
+        }
+    }
+
+    // MARK: - Helpers
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift:287]: Can we organize the methods according to these `MARK: -`s?

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+    }
+
+    private var didSendLogShow: Bool = false
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        if !didSendLogShow {
+            // Only send this once to match the behavior of horizontal mode
+            didSendLogShow = true
+            analyticsHelper.logShow(showingSavedPMList: false) // We never show the saved PM list first
+        }
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift:284]: Per https://docs.google.com/document/d/1hBpsyWQi2d_gGMkS-_aUGUZVn1g53mjlLGytyxraqnU/edit?tab=t.0#heading=h.gnypjvjc9wbz we should delete this; we don't call `logShow`.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+                    // We're done!
+                    let delay: TimeInterval =
+                    self.presentedViewController?.isBeingDismissed == true ? 1 : 0
+                    // Hack: PaymentHandler calls the completion block while SafariVC is still being dismissed - "wait" until it's finished before updating UI
+                    DispatchQueue.main.asyncAfter(deadline: .now() + delay) {
+#if !canImport(CompositorServices)
+                        UINotificationFeedbackGenerator().notificationOccurred(.success)
+#endif
+                        self.primaryButton.update(state: .succeeded, animated: true) {
+                            // Wait a bit before closing the sheet
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift:349]: lol if you do `git log -S "Wait a bit before closing the sheet"` you'll see this was accidentally copied here and in vertical from the original mc implementation which actually added a delay. This comment doesn't make sense now, delete (here and in vertical too please!)

**Review by yuki-stripe — COMMENTED** [review]
Love the clear & simple scope of this PR!

File: StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/EmbeddedFormViewControllerSnapshotTests.swift [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+            formSheetAction: .continue
+        )
+        let sut = makeEmbeddedFormViewController(
+            configuration: configuration,
+            paymentMethodType: .card
+        )
+        verify(sut)
+    }
+
+    func testDisplaysError() {
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/EmbeddedFormViewControllerSnapshotTests.swift:98]: Can we add `testDisplaysErrorAndMandate` too

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+        result: PaymentSheetResult
+    )
+
+    /// Informs the delegate that the user has canceled the payment process.
+    /// This method is called when the user dismisses the view controller or taps the cancel button.
+    /// - Parameter embeddedFormViewController: The view controller that was canceled.
+    func embeddedFormViewControllerDidCancel(_ embeddedFormViewController: EmbeddedFormViewController)
+
+    /// Notifies the delegate that the embedded form view controller should close.
+    /// This method is called when a payment option that can be confirmed later has been provided.
```
[the remark this answers is not in the document]
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift:44]: Yep, that's the continue case, will rename

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+        case .confirm:
+            return false
+        case .continue:
+            return true
+        }
+    }
+
+    var collectsUserInput: Bool {
+        return paymentMethodFormViewController.form.collectsUserInput
+    }
```
[the remark this answers is not in the document]
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift:63]: Yeah, plan is to create an EmbeddedFormViewController when a user taps a row, check if we collect user input to determine if we should present or not. Will remove for now.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+    ///   - embeddedFormViewController: The view controller requesting the confirmation.
+    ///   - paymentOption: The `PaymentOption` to be confirmed.
+    ///   - completion: A completion handler to call with the `PaymentSheetResult` from the confirmation attempt.
+    func embeddedFormViewControllerShouldConfirm(
+        _ embeddedFormViewController: EmbeddedFormViewController,
+        with paymentOption: PaymentOption,
+        completion: @escaping (PaymentSheetResult, STPAnalyticsClient.DeferredIntentConfirmationType?) -> Void
+    )
+
+    /// Informs the delegate that the embedded form view controller has finished and is dismissing with a result.
```
[the remark this answers is not in the document]
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift:29]: Took your doc string suggestion from above so this is no re-worded.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift (hunk 3 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift:284]: Removing!

File: StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/EmbeddedFormViewControllerSnapshotTests.swift (hunk 5 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/EmbeddedFormViewControllerSnapshotTests.swift:98]: Addded!

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift (hunk 1 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift:219]: Hrm, I tried ctrl-m but it formats the file in a really odd way.

```swift
@_spi(
    STP
) import StripeCore
```

And things like

```swift
    func updateErrorLabel(
        for error: Swift.Error?
    ) {
        self.error = error
        updateError()
    }
```

IMO the linter should catch this.

Comment (yuki-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift:219]: Oh yeah I wouldn't run ctrl-m on the whole file.  Just individual code blocks. Agree linter should enforce this

**Review by yuki-stripe — APPROVED** [review]

**Pull request merged**